### PR TITLE
D9 - Set verified checksum contact in the session

### PR DESCRIPTION
**Pull request by jitendrapurohit**
Overview
----------------------------------------
Grant CiviCRM permissions to the verified checksum contact

Before
----------------------------------------
Webform loaded with checksum contact is not able to view related contacts, etc. 

After
----------------------------------------
Fixed.

Technical Details
----------------------------------------
Set verified checksum contact in the session and ensure its removed when loaded anonymously without cs.

Comments
----------------------------------------
@KarinG

### CODE REVIEW

**Comment by KarinG** [comment]
Thank you for this - 

It's failing existing Checksum tests -> 

```
1) Drupal\Tests\webform_civicrm\FunctionalJavascript\LocationTypeTest::testAddressUpdateUsingChecksum
Behat\Mink\Exception\ExpectationException: The field "First Name" value is "", but "Pabst" expected.
```

PS - I'm re-running the tests b/c at least one of them fails on a missing token (that happens sometimes - but then is resolved on re-run).

File: src/Utils.php [hunk 1]
```diff
@@ -1021,4 +1027,81 @@ public function hasMultipleValues($element) {
...
+    $urlChecksum1 = $request->query->get('cs') ?? $request->query->get('cs1') ?? NULL;
+
+    if (!empty($urlChecksum1) && !empty($urlCid1)) {
+      $valid = $this->wf_civicrm_api4('Contact', 'validateChecksum', [
+        'contactId' => $urlCid1,
+        'checksum' => $urlChecksum1,
+      ])[0] ?? [];
+      if ($valid['valid']) {
+        // checkAccess v4 api does not check for access via relationship.
+        if (\CRM_Contact_BAO_Contact_Permission::allow($cid)) {
```
Comment (jitendrapurohit) [inline src/Utils.php:1099]: Currently, the checkAccess API or the contact get api in core does not allow to view the contact if its permitted only by relationship. It only checks for contact allowed via ACL - https://github.com/civicrm/civicrm-core/blob/master/CRM/Contact/BAO/Contact/Permission.php#L361. Raised https://lab.civicrm.org/dev/core/-/issues/4610 for additional thoughts.

Comment (KarinG) [reply src/Utils.php:1099]: Thank you @jitendrapurohit

**Comment by KarinG** [comment]
Hi @jitendrapurohit - I was able to live test this PR on a client site now - but am not able to get things working yet. 

My details:
```
root@be384929ec22:/var/www/drupal/web/modules/contrib/webform_civicrm# cat 904.diff |patch -p1          
patching file src/ContactComponent.php
patching file src/Utils.php
Hunk #2 succeeded at 620 (offset -1 lines).
Hunk #3 succeeded at 1023 (offset -4 lines).
patching file src/UtilsInterface.php
patching file tests/modules/webform_civicrm_test/config/install/webform.webform.update_contact_details.yml
patching file tests/src/FunctionalJavascript/LocationTypeTest.php
Hunk #5 succeeded at 254 (offset 1 line).
patching file tests/src/Unit/UtilsTest.php
patching file webform_civicrm.services.yml
```

drush cr

URL generated from CiviMail -> looks like:
/delegate-registration?cid1=597407&cs=4a7a10a4f94698beaab3e4fa9cc2f370_1701028009_1440

logged in and using
/delegate-registration
all is well
logged out (or logged in) and using the checksum URL:

Error:
`CRM_Core_Exception: Please file an issue with the backtrace in civicrm_api3() (line 135 of /var/www/drupal/vendor/civicrm/civicrm-core/api/api.php).
`
I don't see backtrace in either Configandlog or dblog though...

Key thing about this webform is:
logged in user has a relationship to an organization and the webform is trying to load that ->

This is cid1 -> nothing unusual:
```
    civicrm_1_contact_1_contact_existing:
      '#type': civicrm_contact
      '#title': 'Contact ID'
      '#description': 'Enter Contact ID'
      '#description_display': after
      '#widget': textfield
      '#none_prompt': '+ Create new +'
      '#results_display':
        id: id
      '#default': user
      '#allow_url_autofill': false
      '#check_permissions': 0
      '#allow_create': 1
      '#contact_type': individual
      '#form_key': civicrm_1_contact_1_contact_existing
      '#parent': civicrm_1_contact_1_fieldset_fieldset
      '#extra': {  }
```

This is the organization -> note the #default_relationship

```
civicrm_2_contact_1_fieldset_fieldset:
  '#type': fieldset
  '#title': 'Constituency Association'
  '#form_key': civicrm_2_contact_1_fieldset_fieldset
  civicrm_2_contact_1_contact_existing:
    '#type': civicrm_contact
    '#title': 'Electoral District Association'
    '#widget': hidden
    '#none_prompt': '- None Found -'
    '#show_hidden_contact': 1
    '#results_display':
      display_name: display_name
    '#default': relationship
    '#default_relationship_to': '1'
    '#default_relationship':
      39_b: 39_b
    '#allow_url_autofill': false
    '#filter_relationship_contact': ''
    '#check_permissions': 0
    '#contact_type': organization
    '#form_key': civicrm_2_contact_1_contact_existing
    '#parent': civicrm_2_contact_1_fieldset_fieldset
    '#extra': {  }

```

**Comment by KarinG** [comment]
Actually it IS working now - perhaps another round of cache clears did it. I'm honestly not sure - but the webform is loading the related contact properly now my test contacts - and Dave is confirming that's what he is seeing as well.

**Comment by KarinG** [comment]
Ok that last test fail is a known php 8 issue in iATS master (we're only pulling in iATS master on that combination -> 1) Drupal\Tests\webform_civicrm\FunctionalJavascript\ContributionIatsTest::testSubmitRecurringContribution
Failed asserting that a string is not empty.

**Pull request merged**